src/main/java/au/org/emii/ncdfgenerator/NcdfEncoderBuilder.java

```diff
@@ -0,0 +1,1476 @@
+		// PreparedStatement s = conn.prepareStatement("set search_path=" + schema + ",public");
+		s.execute();
+		s.close();
+
+		String selection = translate.process( selection_expr);
+
+		String query = "SELECT distinct data.instance_id  FROM (" + definition.virtualDataTable + ") as data where " + selection + ";" ;
+		System.out.println( "first query " + query  );
+
+		PreparedStatement stmt = conn.prepareStatement( query );
+		stmt.setFetchSize(fetchSize);
+
+		// try ...
+		// change name featureInstancesRSToProcess ?
+		featureInstancesRS = stmt.executeQuery();
+		System.out.println( "done determining feature instances " );
+		// should determine our target types here
+	}
+
+	public void populateValues(
+		Map< String, IDimension> dimensions,
+		Map< String, IVariableEncoder> encoders,
+		String query
+		)  throws Exception
+	{
+		System.out.println( "query " + query  );
+
+		// sql stuff
+		PreparedStatement stmt = conn.prepareStatement( query );
+		stmt.setFetchSize(fetchSize);
+		ResultSet rs = stmt.executeQuery();
+
+		// now we loop the main attributes
+		ResultSetMetaData m = rs.getMetaData();
+		int numColumns = m.getColumnCount();
+
+		// pre-map the encoders by index according to the column name
+		ArrayList< IAddValue> [] processing = (ArrayList< IAddValue> []) new ArrayList [numColumns + 1];
+
+		for ( int i = 1 ; i <= numColumns ; i++ ) {
+			// System.out.println( "column name "+ m.getColumnName(i) );
+			processing[i] = new ArrayList< IAddValue> ();
+
+			IDimension dimension = dimensions.get( m.getColumnName(i));
+			if( dimension != null)
+				processing[i].add( dimension );
+
+			IAddValue encoder = encoders.get(m.getColumnName(i));
+			if( encoder != null)
+				processing[i].add( encoder );
+		}
+
+		// process result set rows
+		while ( rs.next() ) {
+			for ( int i = 1 ; i <= numColumns ; i++ ) {
+				for( IAddValue p : processing[ i] ) {
+					p.addValueToBuffer( rs.getObject( i));
+				}
+			}
+		}
+	}
+
+
+	public NetcdfFileWriteable get() throws Exception
+	{
+		// TODO should just return a readable IStream, client shouldn't care that it's netcdf type.
+
+		try {
+			if( featureInstancesRS.next())
+			{
+				// munge
+				long instance_id = -1234;
+				Object o = featureInstancesRS.getObject(1);
+				Class clazz = o.getClass();
+				if( clazz.equals( Integer.class )) {
+					instance_id = (long)(Integer)o;
+				}
+				else if( clazz.equals( Long.class )) {
+					instance_id = (long)(Long)o;
+				} else {
+					throw new NcdfGeneratorException( "Can't convert intance_id type to integer" );
+				}
+
+				System.out.println( "whoot get(), instance_id is " + instance_id );
+
+				String selection = translate.process( selection_expr); // we ought to be caching the specific query ???
+
+				populateValues( definition.dimensions, definition.encoders,
+					"SELECT * FROM (" + definition.virtualInstanceTable + ") as instance where instance.id = " + Long.toString( instance_id) );
```

Comment:
I'll think about this. The code is not performance sensitive (which would imply using a string builder), and I personally find + to be quite readable. I agree more needs to be a lot more done, especially regarding parameter handling, since it's all handled as strings at the moment (open to sql injection attacks etc)>